### PR TITLE
ddl: support modify column

### PR DESCRIPTION
**Pull request by coocood**
Currently, only supports column type changes that do not require data check or data change.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala PTAL

**Comment by zimulala** [comment]
Add tests in parse_test.go

File: ddl/ddl.go [hunk 1]
```diff
@@ -431,19 +432,29 @@ func (d *ddl) buildColumnsAndConstraints(ctx context.Context, colDefs []*ast.Col
 	return cols, constraints, nil
 }
 
-func (d *ddl) buildColumnAndConstraint(ctx context.Context, offset int,
-	colDef *ast.ColumnDef) (*table.Column, []*ast.Constraint, error) {
-	// Set charset.
-	if len(colDef.Tp.Charset) == 0 {
-		switch colDef.Tp.Tp {
+func (d *ddl) setCharsetFlenDecimal(tp *types.FieldType) {
```
Comment (zimulala) [inline ddl/ddl.go]: How about "setFieldType" ?

File: ddl/ddl.go [hunk 2]
```diff
@@ -1054,6 +1059,84 @@ func (d *ddl) DropColumn(ctx context.Context, ti ast.Ident, colName model.CIStr)
...
+		return errors.Trace(infoschema.ErrDatabaseNotExists)
+	}
+	t, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists)
+	}
+	colName := spec.Column.Name.Name
+	col := table.FindCol(t.Cols(), colName.L)
+	if col == nil {
+		return infoschema.ErrTableNotExists.Gen("column %s doesn't exist", colName.O)
```
Comment (zimulala) [inline ddl/ddl.go]: Is the error wrong?

**Comment by hanfei1991** [comment]
rest LGTM

File: ddl/ddl.go (hunk 1 above)
Comment (coocood) [inline ddl/ddl.go]: It only sets Flen and Decimal, charset and collation is not set in this function.

**Comment by coocood** [comment]
@zimulala @hanfei1991 PTAL

**Comment by hanfei1991** [comment]
LGTM

File: ddl/column.go [hunk 3]
```diff
@@ -417,3 +417,40 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, handle
...
+	err = t.UpdateTable(job.SchemaID, tblInfo)
+	if err != nil {
+		job.State = model.JobCancelled
+		return errors.Trace(err)
+	}
+	ver, err := updateSchemaVersion(t, job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	job.State = model.JobDone
```
Comment (zimulala) [inline ddl/column.go]: Does here be added job.SchemaState ?

File: ddl/column.go [hunk 4]
```diff
@@ -417,3 +417,40 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, handle
...
+		return errors.Trace(err)
+	}
+	newCol := &model.ColumnInfo{}
+	err = job.DecodeArgs(newCol)
+	if err != nil {
+		job.State = model.JobCancelled
+		return errors.Trace(err)
+	}
+	var done bool
+	for i, col := range tblInfo.Columns {
```
Comment (zimulala) [inline ddl/column.go]: Use the function of `findCol`.

Comment (coocood) [inline ddl/column.go]: `findCol` returns a *model.ConlumnInfo, but we need to replace that column with new column.

File: ddl/ddl.go [hunk 5]
```diff
@@ -1056,12 +1060,90 @@ func (d *ddl) DropColumn(ctx context.Context, ti ast.Ident, colName model.CIStr)
...
+	case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString,
+		mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
+		switch to.Tp {
+		case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString,
+			mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
+			return true
+		default:
+			return false
+		}
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
```
Comment (zimulala) [inline ddl/ddl.go]: Can `mysql.TypeYear`  be added here?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**